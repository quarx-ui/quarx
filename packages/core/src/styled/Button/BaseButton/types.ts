import { ReactNode } from 'react';
import { BaseProps, ComponentPropsWithHTML } from '@core/types';
import { WithClassesAndStyles } from '@core/styles';
import { ButtonStyleParams } from '@core/src';
import { BaseButtonStyleKeys } from './style';

export interface BaseButtonPropsWithoutHtml extends
    Partial<ButtonStyleParams>,
    WithClassesAndStyles<BaseButtonStyleKeys, ButtonStyleParams>,
    BaseProps<HTMLButtonElement>
{
    /** HTML-тип элемента button */
    buttonType?: 'submit' | 'reset' | 'button';

    /** Текст расположенный в компоненте */
    children?: ReactNode,
}

export type BaseButtonProps = ComponentPropsWithHTML<BaseButtonPropsWithoutHtml, 'button'>;
