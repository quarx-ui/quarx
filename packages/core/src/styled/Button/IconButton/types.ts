import { IconButtonStyleKeys } from '@quarx-ui/core/src/styled/Button/IconButton/style';
import { WithClassesAndStyles } from '@core/styles';
import { BaseButtonProps, ButtonStyleParams } from '@core/src';

export type IconButtonProps = Omit<BaseButtonProps, 'classes'>
& WithClassesAndStyles<IconButtonStyleKeys, ButtonStyleParams>
