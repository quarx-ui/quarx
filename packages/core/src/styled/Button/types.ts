import { QxBorderSize, PaletteColor, PickQxSize } from '@core';

export type ButtonSize = PickQxSize<'xSmall' | 'small' | 'medium' | 'large'>;
export type ButtonType = 'contained' | 'outlined' | 'text';
export type ButtonColor = PaletteColor;
export type ButtonBorderRadius = QxBorderSize;

export interface ButtonStyleParams {
    /** Цветовая палитра компонента. Значения соответствуют токенам объекта `palette.colors` */
    color: ButtonColor,

    /** Размер компонента. Влияет на высоту и размер текста
     * @property xSmall минимальный размер
     * @property small маленький
     * @property medium средний
     * @property large наибольший размер */
    size: ButtonSize,

    /** Скругление компонента. Значения соответствуют токенам объекта `borderRadii` */
    borderRadius: ButtonBorderRadius,

    /** Тип заливки компонента
     * @property contained используется, чтобы усилить акцент на основном действии
     * @property outlined используется для дополнительных действий или в элементах, где не нужен максимальный акцент
     * @property text текстовые кнопки обычно используются для не самых важных действий */
    type: ButtonType,

    /** Анимация загрузки */
    loading: boolean,

    /** Изменяет состояние компонента на активное/неактивное. Отключает обработчики события */
    disabled: boolean,
}
