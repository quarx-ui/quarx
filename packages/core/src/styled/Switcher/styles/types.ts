import { PaletteColor, PickQxSize, Values } from '@core';
import { SWITCHER_POSITION } from '@quarx-ui/core/src/styled/Switcher/constants';

export type SwitcherPositionType = Values<typeof SWITCHER_POSITION>;

export interface SwitcherStyleParams {
    /** Состояние наведения */
    hover: boolean,

    /** Размер компонента */
    size: PickQxSize<'small' | 'medium' | 'large'>,

    /** Цвет компонента */
    color?: PaletteColor

    /** Изменяет состояние компонента на активное/неактивное */
    disabled: boolean,

    /** Отключение возможности фокуса */
    disableFocus: boolean,

    /** Позиция переключателя относительно текста */
    position: SwitcherPositionType,

    /** Состояние компонента */
    checked: boolean,
}
