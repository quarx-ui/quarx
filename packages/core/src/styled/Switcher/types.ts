import { ChangeEventHandler, RefObject } from 'react';
import { BaseProps, ComponentPropsWithHTML, SelectionProps, WithClassesAndStyles } from '@core';
import { SwitcherStyleParams } from '@quarx-ui/core/src/styled/Switcher/styles/types';
import { SwitcherCSSVarKeys, SwitcherStyleKeys } from './styles';

export interface SwitcherPropsWithoutHtml extends
    Partial<SwitcherStyleParams>,
    WithClassesAndStyles<SwitcherStyleKeys, SwitcherStyleParams, SwitcherCSSVarKeys>,
    BaseProps<HTMLLabelElement>
{
    /** Обработчик изменения состояния элемента `input` */
    onChange?: ChangeEventHandler<HTMLInputElement>,

    /** Свойство `name` элемента `input` */
    name?: string

    /** Свойство `value` элемента `input` */
    value?: string,

    /** Объект со свойствами для элемента `input` */
    inputProps?: JSX.IntrinsicElements['input'],

    /** Ссылка к элементу `input` */
    inputRef?: RefObject<HTMLInputElement>,
}

export type SwitcherProps = ComponentPropsWithHTML<SwitcherPropsWithoutHtml, 'label'>

export interface SwitcherSelectionProps extends
    SelectionProps,
    Pick<SwitcherProps, 'checked'>
{
    /** Объект с параметрами для компонента `switcher` */
    switcherProps?: SwitcherProps,
}

export * from './styles/types';
