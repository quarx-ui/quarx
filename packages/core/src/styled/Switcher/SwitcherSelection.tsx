import React, { createRef, FC, forwardRef, MouseEventHandler, useState } from 'react';
import { Selection, Switcher, SwitcherSelectionProps } from '@core';
import { usePropsOverwrites } from '@core/styles';

export const SwitcherSelection: FC<SwitcherSelectionProps> = forwardRef<HTMLDivElement, SwitcherSelectionProps>((
    initialProps,
    ref,
) => {
    const { props, cn, styleProps } = usePropsOverwrites('SwitcherSelection', initialProps);
    const {
        focused: externalFocus,
        switcherProps,
        disabled,
        size,
        onMouseLeave,
        onMouseEnter,
        checked,
    } = props;

    const [focus, setFocus] = useState(externalFocus ?? false);
    const [hover, setHover] = useState(false);

    const mouseLeaveHandler: MouseEventHandler<HTMLDivElement> = (e) => {
        setHover(false);
        onMouseLeave?.(e);
    };

    const mouseEnterHandler: MouseEventHandler<HTMLDivElement> = (e) => {
        setHover(true);
        onMouseEnter?.(e);
    };

    const switcherRef = createRef<HTMLInputElement>();

    return (
        <Selection
            {...props}
            {...styleProps}
            className={cn('root')}
            onMouseEnter={mouseEnterHandler}
            onMouseLeave={mouseLeaveHandler}
            disabled={switcherRef.current?.disabled || disabled}
            focused={externalFocus ?? focus}
            ref={ref}
            leftAdornment={(
                <Switcher
                    inputRef={switcherRef}
                    color={switcherProps?.color}
                    disabled={switcherRef.current?.disabled || disabled}
                    checked={checked}
                    hover={hover}
                    disableFocus
                    size={size}
                    {...switcherProps}
                    onFocus={() => setFocus(true)}
                    onBlur={() => setFocus(false)}
                />
            )}
        />
    );
});
