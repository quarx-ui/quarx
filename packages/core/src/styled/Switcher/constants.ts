import { valuesAsKeysFromArray } from '@core';

export const SWITCHER_POSITION = valuesAsKeysFromArray([
    'right', 'left',
]);
