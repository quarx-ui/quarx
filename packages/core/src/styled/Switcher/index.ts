export * from './types';
export * from './Switcher';
export * from './constants';
