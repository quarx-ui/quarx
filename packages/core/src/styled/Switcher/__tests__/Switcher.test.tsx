import { cleanup, render, screen } from '@testing-library/react';
import { SwitcherProps, SwitcherStyleParams } from '@quarx-ui/core/src/styled/Switcher/types';
import React from 'react';
import { Switcher } from '@core';
import userEvent from '@testing-library/user-event';
import { testStyleParams } from '@core/test-utils';

describe('Switcher', () => {
    testStyleParams<Partial<SwitcherStyleParams>, SwitcherProps>(
        Switcher,
        {
            size: 'medium',
            color: 'brand',
            position: 'left',
        },
        { children: 'Статус' },
    )({
        color: ['brand', 'secondary', 'info', 'success', 'warning', 'danger'],
        size: ['small', 'medium', 'large'],
        position: ['left', 'right'],
        disabled: [true, false],
    });

    it('text should be in the document', () => {
        render(<Switcher>Switcher</Switcher>);

        expect(screen.queryByText('Switcher')).toBeInTheDocument();
        const checkbox = document.querySelector('input');
        expect(checkbox).not.toHaveAttribute('disabled');
    });

    it('Switcher should be checked', () => {
        render(<Switcher checked>Switcher</Switcher>);

        const checkbox = document.querySelector('input');
        expect(checkbox?.checked).toEqual(true);
    });

    it('Switcher should have attribute disabled', () => {
        const mockedOnClick = jest.fn();
        render(<Switcher disabled onClick={mockedOnClick}>Switcher</Switcher>);
        const input = document.querySelector('input');
        const checkbox = document.querySelector('label');

        if (!input || !checkbox) { return; }

        expect(input).toHaveAttribute('disabled');
        userEvent.click(checkbox);
        expect(mockedOnClick).not.toBeCalled();
    });

    it('Switcher should be hidden', () => {
        render(<Switcher hidden>Switcher</Switcher>);
        expect(document.querySelector('input')).not.toBeInTheDocument();

        cleanup();

        render(<Switcher permissions={{ hidden: true }}>Switcher</Switcher>);
        expect(document.querySelector('input')).not.toBeInTheDocument();
    });

    it('onClick should be called', () => {
        let checked = false;
        const mockedOnClick = jest.fn(() => {
            checked = !checked;
        });
        render(<Switcher onChange={mockedOnClick}>Switcher</Switcher>);
        const checkbox = document.querySelector('label');

        if (!checkbox) { return; }

        userEvent.click(checkbox);
        expect(mockedOnClick).toBeCalledTimes(1);
        expect(checked).toEqual(true);
    });
});
