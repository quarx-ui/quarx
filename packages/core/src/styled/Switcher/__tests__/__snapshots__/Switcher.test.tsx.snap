// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher style params color brand 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_position_left css-1vn84au-QxSwitcher-root"
    data-testid="testStyleParams-color-brand"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params color danger 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_danger QxSwitcher_position_left css-kh8wo2-QxSwitcher-root"
    data-testid="testStyleParams-color-danger"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-do69oj-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params color info 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_info QxSwitcher_position_left css-12vrftk-QxSwitcher-root"
    data-testid="testStyleParams-color-info"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-do69oj-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params color secondary 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_secondary QxSwitcher_position_left css-h6m773-QxSwitcher-root"
    data-testid="testStyleParams-color-secondary"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params color success 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_success QxSwitcher_position_left css-1ekzyed-QxSwitcher-root"
    data-testid="testStyleParams-color-success"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-do69oj-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params color warning 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_warning QxSwitcher_position_left css-17m083m-QxSwitcher-root"
    data-testid="testStyleParams-color-warning"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-do69oj-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params disabled false 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_position_left css-1vn84au-QxSwitcher-root"
    data-testid="testStyleParams-disabled-false"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params disabled true 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_disabled QxSwitcher_position_left css-17dqn1m-QxSwitcher-root"
    data-testid="testStyleParams-disabled-true"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      disabled=""
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-1iima7a-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-1u5g9a4-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params position left 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_position_left css-1vn84au-QxSwitcher-root"
    data-testid="testStyleParams-position-left"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params position right 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_position_right css-1vn84au-QxSwitcher-root"
    data-testid="testStyleParams-position-right"
  >
    <div
      class="QxSwitcher-content css-1r6yxqi-QxSwitcher-content"
    >
      Статус
    </div>
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params size large 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_large QxSwitcher_color_brand QxSwitcher_position_left css-o2j576-QxSwitcher-root"
    data-testid="testStyleParams-size-large"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-p24uv7-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params size medium 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_medium QxSwitcher_color_brand QxSwitcher_position_left css-1vn84au-QxSwitcher-root"
    data-testid="testStyleParams-size-medium"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-e6slwr-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;

exports[`Switcher style params size small 1`] = `
<DocumentFragment>
  <label
    class="QxSwitcher QxSwitcher_size_small QxSwitcher_color_brand QxSwitcher_position_left css-1ky9bec-QxSwitcher-root"
    data-testid="testStyleParams-size-small"
  >
    <input
      class="QxSwitcher-input css-vbnokk-QxSwitcher-input"
      type="checkbox"
      value=""
    />
    <div
      class="QxSwitcher-toggleContainer css-iqwm0s-QxSwitcher-toggleContainer"
    >
      <div
        class="QxSwitcher-toggle css-3p3seb-QxSwitcher-toggle"
      />
    </div>
    <div
      class="QxSwitcher-content css-1ywo4bn-QxSwitcher-content"
    >
      Статус
    </div>
  </label>
</DocumentFragment>
`;
